librz/bin/dwarf.c

```diff
@@ -2191,193 +1955,128 @@ static ut8 *get_section_bytes(RzBin *bin, const char *sect_name, size_t *len) {
  *
  * @param da Parsed abbreviations
  * @param bin
- * @param mode RZ_MODE_PRINT to print
  * @return RzBinDwarfDebugInfo* Parsed information, NULL if error
  */
-RZ_API RzBinDwarfDebugInfo *rz_bin_dwarf_parse_info(RzBinDwarfDebugAbbrev *da, RzBin *bin, int mode) {
-	RzBinDwarfDebugInfo *info = NULL;
-	RzBinSection *debug_str;
-	RzBinSection *section = getsection(bin, "debug_info");
-	RzBinFile *binfile = bin ? bin->cur : NULL;
+RZ_API RzBinDwarfDebugInfo *rz_bin_dwarf_parse_info(RzBinFile *binfile, RzBinDwarfDebugAbbrev *da) {
+	rz_return_val_if_fail(binfile, NULL);
+	RzBinSection *section = getsection(binfile, "debug_info");
+	if (!section) {
+		return NULL;
+	}
 
+	RzBinDwarfDebugInfo *info = NULL;
 	ut64 debug_str_len = 0;
 	ut8 *debug_str_buf = NULL;
 
-	if (binfile && section) {
-		debug_str = getsection(bin, "debug_str");
-		if (debug_str) {
-			debug_str_len = debug_str->size;
-			debug_str_buf = calloc(1, debug_str_len + 1);
-			if (!debug_str_buf) {
-				goto cleanup;
-			}
-			st64 ret = rz_buf_read_at(binfile->buf, debug_str->paddr,
-				debug_str_buf, debug_str_len);
-			if (!ret) {
-				goto cleanup;
-			}
-		}
-
-		ut64 len = section->size;
-		// what is this checking for?
-		if (len > (UT32_MAX >> 1) || len < 1) {
-			goto cleanup;
-		}
-		ut8 *buf = calloc(1, len);
-		if (!buf) {
-			goto cleanup;
-		}
-		if (!rz_buf_read_at(binfile->buf, section->paddr, buf, len)) {
-			free(buf);
-			goto cleanup;
+	RzBinSection *debug_str = debug_str = getsection(binfile, "debug_str");
+	if (debug_str) {
+		debug_str_len = debug_str->size;
+		debug_str_buf = calloc(1, debug_str_len + 1);
+		if (!debug_str_buf) {
+			goto cave;
 		}
-		/* set the endianity global [HOTFIX] */
-		big_end = rz_bin_is_big_endian(bin);
-		info = parse_info_raw(binfile->sdb_addrinfo, da, buf, len,
+		st64 ret = rz_buf_read_at(binfile->buf, debug_str->paddr,
 			debug_str_buf, debug_str_len);
+		if (!ret) {
+			goto cave_debug_str_buf;
+		}
+	}
 
-		if (mode == RZ_MODE_PRINT && info) {
-			print_debug_info(info, bin->cb_printf);
-		}
-		// build hashtable after whole parsing because of possible relocations
-		if (info) {
-			size_t i, j;
-			for (i = 0; i < info->count; i++) {
-				RzBinDwarfCompUnit *unit = &info->comp_units[i];
-				for (j = 0; j < unit->count; j++) {
-					RzBinDwarfDie *die = &unit->dies[j];
-					ht_up_insert(info->lookup_table, die->offset, die); // optimization for further processing}
-				}
+	ut64 len = section->size;
+	// what is this checking for?
+	if (len > (UT32_MAX >> 1) || len < 1) {
```

Comment:
(answering to the comment) sanitize input, i guess?

Reply:
The comment was from old code, I think it was referring specifically to the `UT32_MAX` check which I don't understand either. I'll remove that and only check for `len` since that makes sense.